### PR TITLE
Use binding name/namespace as infrastructure secret identifier

### PR DESCRIPTION
**Pull request by petersutter**
**What this PR does / why we need it**:
Re-introduce name and namespace for infrastructure secrets by using bindingName and bindingSecret.
This also fixes secret related bugs introduced with https://github.com/gardener/dashboard/pull/860.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement user

```

### CODE REVIEW

File: backend/lib/services/infrastructureSecrets.js [hunk 1]
```diff
@@ -248,27 +250,33 @@ exports.patch = async function ({ user, namespace, bindingName, body }) {
...
   const client = user.client
 
-  const secretBinding = await client['core.gardener.cloud'].secretbindings.get(namespace, bindingName)
+  const secretBinding = await client['core.gardener.cloud'].secretbindings.get(namespace, name)
   const secretName = _.get(secretBinding, 'secretRef.name')
+  const secretNamespace = _.get(secretBinding, 'secretRef.namespace')
 
-  checkIfOwnSecret(secretBinding)
+  if (!isOwnSecret(secretBinding)) {
+    throw new MethodNotAllowed('Remove allowed only for secrets in own namespace')
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:260]: https://tools.ietf.org/html/rfc7231#section-6.5.5 
Use 422

File: backend/lib/services/infrastructureSecrets.js [hunk 2]
```diff
@@ -248,27 +250,33 @@ exports.patch = async function ({ user, namespace, bindingName, body }) {
...
+  }
 
   const { items: shootList } = await shoots.list({ user, namespace })
-  const predicate = (item) => {
-    const itemSecretBindingName = _.get(item, 'spec.secretBindingName')
-    return itemSecretBindingName === bindingName
-  }
-  const secretReferencedByShoot = _.find(shootList, predicate)
+  const secretReferencedByShoot = _.find(shootList, ['spec.secretBindingName', name])
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:264]: Not part of this PR but also PreconditionFailed is not used correctly here. All of these erros should be 422.

File: backend/lib/services/infrastructureSecrets.js [hunk 3]
```diff
@@ -248,27 +250,33 @@ exports.patch = async function ({ user, namespace, bindingName, body }) {
...
+    await client['core.gardener.cloud'].secretbindings.delete(namespace, name),
+    await client.core.secrets.delete(secretNamespace, secretName)
   ])
-  return { metadata: { name: secretName, bindingName, namespace } }
+  return {
+    metadata: {
+      name,
+      namespace,
+      secretName,
+      secretNamespace
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js]: should be secretRef I think

File: backend/test/acceptance/api.infrastructureSecrets.spec.js [hunk 4]
```diff
@@ -119,34 +119,34 @@ module.exports = function ({ agent, sandbox, k8s, auth }) {
...
+    k8s.stub.deleteInfrastructureSecret({ bearer, name, namespace, project, secretName, secretNamespace: namespace, cloudProfileName, resourceVersion })
     const res = await agent
-      .delete(`/api/namespaces/${namespace}/infrastructure-secrets/${bindingName}`)
+      .delete(`/api/namespaces/${namespace}/infrastructure-secrets/${name}`)
       .set('cookie', await user.cookie)
 
     expect(res).to.have.status(200)
     expect(res).to.be.json
-    expect(res.body.metadata).to.eql({ name, bindingName, namespace })
+    expect(res.body.metadata).to.eql({ secretName, name, namespace })
```
Comment (holgerkoser) [inline backend/test/acceptance/api.infrastructureSecrets.spec.js]: secretref

**Review by grolu — APPROVED** [review]
/lgtm

File: backend/lib/routes/infrastructureSecrets.js [hunk 5]
```diff
@@ -39,9 +39,9 @@ router.route('/:name')
     try {
       const user = req.user
       const namespace = req.params.namespace
-      const bindingName = req.params.name
+      const name = req.params.name
```
Comment (holgerkoser) [inline backend/lib/routes/infrastructureSecrets.js:42]:
```
const { namespace, name } = req.params
```

File: backend/lib/routes/infrastructureSecrets.js [hunk 6]
```diff
@@ -50,8 +50,8 @@ router.route('/:name')
     try {
       const user = req.user
       const namespace = req.params.namespace
-      const bindingName = req.params.name
-      res.send(await infrastructureSecrets.remove({ user, namespace, bindingName }))
+      const name = req.params.name
```
Comment (holgerkoser) [inline backend/lib/routes/infrastructureSecrets.js:53]:
```
const { namespace, name } = req.params
```

File: backend/lib/services/infrastructureSecrets.js [hunk 7]
```diff
@@ -20,12 +20,14 @@ function fromResource ({ secretBinding, cloudProviderKind, secret, quotas = [],
 
   const infrastructureSecret = {
     metadata: {
-      secretName: _.get(secretBinding, 'secretRef.name'),
-      secretNamespace: _.get(secretBinding, 'secretRef.namespace'),
+      namespace: _.get(secretBinding, 'metadata.namespace'),
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:23]:
```
  const { namespace, name } = secretBinding.metadata
  const secretRef = secretBinding.secretRef
```

File: backend/lib/services/infrastructureSecrets.js [hunk 8]
```diff
@@ -72,10 +74,10 @@ function toSecretResource ({ metadata, data }) {
...
   const kind = resource.kind
   const type = 'Opaque'
-  metadata = _
-    .chain(metadata)
-    .pick(['namespace', 'name'])
-    .value()
+  metadata = {
+    name: metadata.secretRef.name,
+    namespace: metadata.secretRef.namespace
+  }
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:80]:
```suggestion
  metadata = { ...metadata.secretRef }
```

File: backend/lib/services/infrastructureSecrets.js [hunk 9]
```diff
@@ -88,11 +90,8 @@ function toSecretBindingResource ({ metadata }) {
   const resource = Resources.SecretBinding
   const apiVersion = resource.apiVersion
   const kind = resource.kind
-  const name = metadata.bindingName
-  const secretRef = {
-    name: metadata.name,
-    namespace: metadata.namespace
-  }
+  const name = metadata.name
+  const secretRef = metadata.secretRef
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:94]:
```suggestion
  const { name, secretRef } = metadata
```

File: backend/lib/services/infrastructureSecrets.js [hunk 10]
```diff
@@ -216,28 +216,29 @@ exports.create = async function ({ user, namespace, body }) {
   })
 }
 
-function checkIfOwnSecret (bodySecretBinding) {
-  const secretNamespace = _.get(bodySecretBinding, 'secretRef.namespace')
-  const secretBindingNamespace = _.get(bodySecretBinding, 'metadata.namespace')
+function isOwnSecret (secretBinding) {
+  const secretNamespace = _.get(secretBinding, 'secretRef.namespace')
+  const namespace = _.get(secretBinding, 'metadata.namespace')
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:221]:
```suggestion
function isOwnSecret ({ metadata, secretRef }) {
```

File: backend/lib/services/infrastructureSecrets.js [hunk 11]
```diff
@@ -216,28 +216,29 @@ exports.create = async function ({ user, namespace, body }) {
...
 
-  if (secretNamespace !== secretBindingNamespace) {
-    throw new MethodNotAllowed('Patch allowed only for secrets in own namespace')
-  }
+  return secretNamespace === namespace
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:223]:
```suggestion
  return secretRef.namespace === metadata.namespace
```

File: backend/lib/services/infrastructureSecrets.js [hunk 12]
```diff
@@ -216,28 +216,29 @@ exports.create = async function ({ user, namespace, body }) {
...
 }
 
-exports.patch = async function ({ user, namespace, bindingName, body }) {
+exports.patch = async function ({ user, namespace, name, body }) {
   const client = user.client
 
-  const secretBinding = await client['core.gardener.cloud'].secretbindings.get(namespace, bindingName)
+  const secretBinding = await client['core.gardener.cloud'].secretbindings.get(namespace, name)
   const secretName = _.get(secretBinding, 'secretRef.name')
+  const secretNamespace = _.get(secretBinding, 'secretRef.namespace')
```
Comment (holgerkoser) [inline backend/lib/services/infrastructureSecrets.js:231]: We could also here use `secretRef.namespace` instead of `secretNamespace`

File: backend/test/acceptance/api.shoots.spec.js [hunk 13]
```diff
@@ -36,7 +36,7 @@ module.exports = function ({ agent, sandbox, k8s, auth }) {
     kind,
     profile,
     region,
-    bindingName: secret,
+    secretBindingName: secret,
```
Comment (holgerkoser) [inline backend/test/acceptance/api.shoots.spec.js:39]: rename `secret` to `secretBindingName`

File: frontend/src/components/Secret.vue [hunk 14]
```diff
@@ -61,7 +61,7 @@ SPDX-License-Identifier: Apache-2.0
 import { mapGetters } from 'vuex'
 import SecretRow from '@/components/SecretRow'
 import InfraIcon from '@/components/VendorIcon'
-import { isOwnSecretBinding } from '@/utils'
+import { isOwnSecret } from '@/utils'
```
Comment (holgerkoser) [inline frontend/src/components/Secret.vue:64]: A mixin for Secret could make sense

File: frontend/src/components/dialogs/SecretDialog.vue [hunk 15]
```diff
@@ -221,14 +214,8 @@ export default {
...
     shootsByInfrastructureSecret () {
-      const secretName = get(this.secret, 'metadata.name')
-      if (secretName) {
-        const predicate = item => {
-          return get(item, 'spec.secretBindingName') === secretName
-        }
-        return filter(this.shootList, predicate)
-      }
-      return []
+      const name = get(this.secret, 'metadata.name')
```
Comment (holgerkoser) [inline frontend/src/components/dialogs/SecretDialog.vue:217]: I know the following is not part of this PR but shouldn't we rename `InfrastructureSecret` to `InfrastructureSecretBinding` or `InfrastructureBinding` or `InfrastructureCredentials` to clearly distinguish between secret and the thing that keeps the ref to the secret?

File: frontend/src/store/modules/infrastructureSecrets.js [hunk 16]
```diff
@@ -21,8 +21,8 @@ const state = {
 
 // getters
 const getters = {
-  getInfrastructureSecretByBindingName: (state) => ({ name, namespace }) => {
-    return find(state.all, eqlBindingNameAndNamespace({ bindingName: name, bindingNamespace: namespace }))
+  getInfrastructureSecretByName: (state) => ({ name, namespace }) => {
+    return find(state.all, eqlNameAndNamespace({ name, namespace }))
```
Comment (holgerkoser) [inline frontend/src/store/modules/infrastructureSecrets.js:25]:
```suggestion
  getInfrastructureSecretByName (state) {
    return ({ name, namespace }) => find(state.all, eqlNameAndNamespace({ name, namespace }))
  }
```

File: frontend/src/store/modules/infrastructureSecrets.js [hunk 17]
```diff
@@ -42,8 +42,8 @@ const actions = {
   },
   update: ({ commit, rootState }, { metadata, data }) => {
     const namespace = metadata.namespace || rootState.namespace
-    const bindingName = metadata.bindingName
-    return updateInfrastructureSecret({ namespace, bindingName, data: { metadata, data } })
+    const name = metadata.name
```
Comment (holgerkoser) [inline frontend/src/store/modules/infrastructureSecrets.js:45]:
```
const { namespace = rootState.namespace, name } = metadata
```

File: backend/lib/routes/infrastructureSecrets.js (hunk 5 above)
Comment (petersutter) [reply backend/lib/routes/infrastructureSecrets.js:42]: ✅

File: backend/lib/routes/infrastructureSecrets.js (hunk 6 above)
Comment (petersutter) [reply backend/lib/routes/infrastructureSecrets.js:53]: ✅

File: backend/lib/services/infrastructureSecrets.js (hunk 7 above)
Comment (petersutter) [reply backend/lib/services/infrastructureSecrets.js:23]: ✅

File: backend/lib/services/infrastructureSecrets.js (hunk 12 above)
Comment (petersutter) [reply backend/lib/services/infrastructureSecrets.js:231]: ✅

File: backend/lib/services/infrastructureSecrets.js (hunk 1 above)
Comment (petersutter) [reply backend/lib/services/infrastructureSecrets.js:260]: ✅

File: backend/lib/services/infrastructureSecrets.js (hunk 2 above)
Comment (petersutter) [reply backend/lib/services/infrastructureSecrets.js:264]: ✅

File: backend/test/acceptance/api.shoots.spec.js (hunk 13 above)
Comment (petersutter) [reply backend/test/acceptance/api.shoots.spec.js:39]: ✅

File: frontend/src/store/modules/infrastructureSecrets.js (hunk 17 above)
Comment (petersutter) [reply frontend/src/store/modules/infrastructureSecrets.js:45]: ✅

File: frontend/src/utils/index.js [hunk 18]
```diff
@@ -288,8 +288,8 @@ export function getTimeStringTo (time, toTime, withoutPrefix = false) {
   }
 }
 
-export function isOwnSecretBinding (secret) {
-  return get(secret, 'metadata.secretNamespace') === get(secret, 'metadata.bindingNamespace')
+export function isOwnSecret (secret) {
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/utils/index.js:291]: ✅

**Review by holgerkoser — APPROVED** [review]
/lgtm
